### PR TITLE
Add a conversion of individual operations in FQ2I pass.

### PR DESCRIPTION
**Pull request by Icemist**

### CODE REVIEW

**Comment by masahi** [comment]
Can you show some accuracy results on QAT models (BERT etc)?

**Comment by masahi** [comment]
Please also show the reference result (accuracy without fq2i).

**Comment by Icemist** [comment]
> Please also show the reference result (accuracy without fq2i).

| Model type       | Accuracy                                                    |
| ---------------- | ----------------------------------------------------------- |
| without old fq2i | {"exact_match": 82.41248817407758, "f1": 90.06966006174216} |
| with old fq2i    | {"exact_match": 82.41248817407758, "f1": 90.06966006174216} |
| with new fq2i    | {"exact_match": 82.73415326395458, "f1": 90.17821060674615} |

File: src/relay/transforms/fake_quantization_to_integer.cc [hunk 1]
```diff
@@ -270,8 +293,233 @@ class FakeQuantizationRewriter : public MixedModeMutator {
   const bool hard_fail_;
 };
 
+bool is_op_enabled_for_optional_fq2i(const CallNode* call_node) {
```
Comment (mbrookhart) [inline src/relay/transforms/fake_quantization_to_integer.cc:296]: Could we add some comments and advice about what ops should be included in this list?

File: src/relay/transforms/fake_quantization_to_integer.cc [hunk 2]
```diff
@@ -270,8 +293,233 @@ class FakeQuantizationRewriter : public MixedModeMutator {
...
+  const bool hard_fail_;
+};
+
 Expr FakeQuantizationToInteger(const Expr& expr, const IRModule& mod, bool hard_fail) {
-  return FakeQuantizationRewriter(hard_fail).Mutate(expr);
+  auto fq_expr = FakeQuantizationRewriter(hard_fail).Mutate(expr);
+  auto fq_inferred_expr = tvm::relay::InferType(fq_expr);
+  auto ofq_expr = OptionalFakeQuantizationRewriter(hard_fail).Mutate(fq_inferred_expr);
+  return ofq_expr;
 }
```
Comment (mbrookhart) [inline src/relay/transforms/fake_quantization_to_integer.cc:523]: I'm not sure how problematic this will be in non-QAT models, but would it make sense to add another bool to make the "Optional" part of the pass actually optional?

**Review by mbrookhart — CHANGES_REQUESTED** [review]
Overall very nice, only comments are around improved documentation and an optional flag.

@masahi and @elvin-n and I talked a couple of months ago and agreed that the QAT version of the pass (which you're adding here) should probably be a separate pass from the explicit fake quantized model ala tflite and TensorRT. I go back and forth on what the right final form is (I've found edge cases for both versions), but I'd love some improved comments on why this particular design

File: tests/python/relay/test_pass_fake_quantization_to_integer.py [hunk 3]
```diff
@@ -627,3 +628,217 @@ def conv2d(expr, type_map):  # pylint: disable=unused-variable
...
+    print("result_exp\n", result_exp)
+    if allow_rounding_error:
+        assert np.all(np.abs(result_def.astype("int32") - result_int.astype("int32")) <= 1)
+    else:
+        assert np.array_equal(result_def, result_int)
+
+    assert np.array_equal(result_int, result_exp)
+
+
+def test_fq2i_optional_op_chaind_with_disabled_op():
```
Comment (AndrewZhaoLuo) [inline tests/python/relay/test_pass_fake_quantization_to_integer.py:669]: Some of these tests are bit complicated, so I would ask a short comment explaining what this tests for

It'll also help me review

Comment (AndrewZhaoLuo) [inline tests/python/relay/test_pass_fake_quantization_to_integer.py:669]: `chand` --> `chained`

File: tests/python/relay/test_pass_fake_quantization_to_integer.py [hunk 4]
```diff
@@ -627,3 +628,217 @@ def conv2d(expr, type_map):  # pylint: disable=unused-variable
...
+    op0 = relay.qnn.op.qnn.batch_matmul(
+        a, b, relay.const(0), relay.const(0), relay.const(2.0), relay.const(6.0)
+    )
+    op1 = relay.qnn.op.qnn.dequantize(op0, relay.const(12.0), relay.const(0))
+    op2 = relay.op.add(op1, relay.const(1.0))
+    expected_expr = relay.op.erf(op2)
+
+    x_np = np.random.randint(-128, 127, size=shape_x, dtype="int8")
+    w_np = np.random.randint(-128, 127, size=shape_w, dtype="int8")
+    compare_expected_fq_to_int(expr, expected_expr, [x_np, w_np], False)
```
Comment (AndrewZhaoLuo) [inline tests/python/relay/test_pass_fake_quantization_to_integer.py:690]: nit: don't need to include default_kwargs (allow_rounding=False)

**Review by AndrewZhaoLuo — COMMENTED** [review]
Need to read pass a little more carefully to understand difference between this and existing fq2i passes

**Comment by masahi** [comment]
> Need to read pass a little more carefully to understand difference between this and existing fq2i passes

The key is that in the models that existing pass expects, there is always matching dq and q. This PR handles cases such assumption doesnt hold.

**Comment by AndrewZhaoLuo** [comment]
That is understood, my main concern is a lot of code looks similar and I'm wondering if this "optional" pass can supersede the original pass.

**Comment by elvin-n** [comment]
> That is understood, my main concern is a lot of code looks similar

Ideologically transformations look similar, but it is unable to extend current transformation exactly due to its expectation of DQ->dubgraph->Q pattern. And it is not a lot of duplication - all op conversion defined through `FTVMFakeQuantizationToInteger` mechanism is reused, that is great

> I'm wondering if this "optional" pass can supersede the original pass.

Interesting question. Probably it can work in that way as well, but should not be so important.

File: src/relay/transforms/fake_quantization_to_integer.cc (hunk 2 above)
Comment (elvin-n) [reply src/relay/transforms/fake_quantization_to_integer.cc:523]: let's add parameter `enableQAT` to the FakeQuantizationToInteger pass with default value not to call QAT transformation

File: src/relay/transforms/fake_quantization_to_integer.cc (hunk 1 above)
Comment (Icemist) [reply src/relay/transforms/fake_quantization_to_integer.cc:308]: I added a comment of how I selected these operations.

File: src/relay/transforms/fake_quantization_to_integer.cc (hunk 2 above)
Comment (Icemist) [reply src/relay/transforms/fake_quantization_to_integer.cc]: I added the parameter, but called it use_qat. I'm not sure if CamelCase fits here, and I've seen qat in other libraries shortened to lower case.

File: src/relay/transforms/fake_quantization_to_integer.cc [hunk 5]
```diff
@@ -270,8 +293,233 @@ class FakeQuantizationRewriter : public MixedModeMutator {
   const bool hard_fail_;
 };
 
+bool is_op_enabled_for_optional_fq2i(const CallNode* call_node) {
+  const Op op = Downcast<Op>(call_node->op);
+  static auto fqfq = Op::GetAttrMap<FTVMFakeQuantizationToInteger>("FTVMFakeQuantizationToInteger");
+  static std::unordered_set<Op, tvm::ObjectHash, tvm::ObjectEqual> ops = {
+      Op::Get("reshape"),
```
[the remark this answers is not in the document]
Comment (Icemist) [reply src/relay/transforms/fake_quantization_to_integer.cc]: I applied the alphabetical order. Originally they were grouped by groups or pairs.

File: include/tvm/relay/qnn/op/dequantize.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+/*!
+ * \file tvm/relay/executor.h
+ * \brief Relay dequantize.
+ */
+#ifndef TVM_RELAY_QNN_OP_DEQUANTIZE_H_
+#define TVM_RELAY_QNN_OP_DEQUANTIZE_H_
+
+#include <tvm/relay/expr.h>
```
[the remark this answers is not in the document]
Comment (Icemist) [reply include/tvm/relay/qnn/op/dequantize.h]: I preferred to delete it and use the declaration you added with the patch a day ago.

File: src/relay/transforms/type_infer.cc [hunk 7]
```diff
@@ -918,11 +919,7 @@ Type InferTypeLocal(const Expr& expr) {
   mod = transform::InferType()(mod);
 
   Type result_type;
-  if (expr.as<FunctionNode>()) {
-    result_type = mod->Lookup("main")->checked_type();
-  } else {
-    result_type = mod->Lookup("main").as<FunctionNode>()->body->checked_type();
-  }
+  result_type = relay::InferType(sub_graph)->checked_type();
```
[the remark this answers is not in the document]
Comment (Icemist) [reply src/relay/transforms/type_infer.cc:920]: Good catch, I really thought I took care of that.

File: tests/python/relay/test_pass_fake_quantization_to_integer.py [hunk 8]
```diff
@@ -627,3 +628,217 @@ def conv2d(expr, type_map):  # pylint: disable=unused-variable
...
     with pytest.raises(Exception):
         mod_int = tvm.relay.transform.FakeQuantizationToInteger(hard_fail=True)(mod)
+
+
+def compare_expected_fq_to_int(expr, expected_expr, args, allow_rounding_error=False):
+    mod = tvm.IRModule.from_expr(expr)
+    mod_def = tvm.relay.transform.InferType()(mod)
+    mod_int = tvm.relay.transform.FakeQuantizationToInteger(False)(mod_def)
+    mod_exp = tvm.relay.transform.InferType()(tvm.IRModule.from_expr(expected_expr))
+    print("mod_def\n", mod_def, "\n")
```
[the remark this answers is not in the document]
Comment (Icemist) [reply tests/python/relay/test_pass_fake_quantization_to_integer.py]: Deleted.

File: tests/python/relay/test_pass_fake_quantization_to_integer.py (hunk 3 above)
Comment (Icemist) [reply tests/python/relay/test_pass_fake_quantization_to_integer.py]: I tried to explain the motivation by updating the titles and adding a line of description to the tests.

Comment (Icemist) [reply tests/python/relay/test_pass_fake_quantization_to_integer.py]: Renamed the function in another way.

**Comment by Icemist** [comment]
> That is understood, my main concern is a lot of code looks similar and I'm wondering if this "optional" pass can supersede the original pass.

I guess this is a slightly different approach.  Not all operations are suitable for the new pass.
Some operations like "add" require an explicit out_scale, which in the existing pass is taken from the quantization operation. 
This is possible for the existing pass because it goes from the end to the beginning of the selected subgraph. 
In the new pass we go from the beginning to the end. Starting from the dequantization operations and have no access to the quantization operation info at the moments of converting. Or maybe even stop halfway before we get the info.

File: src/relay/transforms/fake_quantization_to_integer.cc [hunk 9]
```diff
@@ -270,16 +297,252 @@ class FakeQuantizationRewriter : public MixedModeMutator {
...
+      Op::Get("nn.pad"),
+      Op::Get("nn.relu"),
+      Op::Get("reshape"),
+      Op::Get("split"),
+      Op::Get("squeeze"),
+      Op::Get("strided_slice"),
+      Op::Get("transpose")};
+
+  auto is_enabled = [&](const auto i) { return i == call_node->op; };
+  auto result = std::find_if(std::begin(ops), std::end(ops), is_enabled);
```
Comment (AndrewZhaoLuo) [inline src/relay/transforms/fake_quantization_to_integer.cc:338]: can you just use `ops.find(...)` from unordered_set? find_if is O(n)

**Review by AndrewZhaoLuo — APPROVED** [review]
LGTM, one more small q

File: src/relay/transforms/fake_quantization_to_integer.cc (hunk 9 above)
Comment (Icemist) [reply src/relay/transforms/fake_quantization_to_integer.cc]: done.

**Review by mbrookhart — APPROVED** [review]
